File: frontend/src/utils/numberToWords.js
export const numberToWords = (num) => {
  if (num === 0) return "Zero";
  
  const a = [
    "", "One", "Two", "Three", "Four", "Five", "Six", "Seven", "Eight", "Nine", "Ten",
    "Eleven", "Twelve", "Thirteen", "Fourteen", "Fifteen", "Sixteen", "Seventeen", "Eighteen", "Nineteen"
  ];
  const b = ["", "", "Twenty", "Thirty", "Forty", "Fifty", "Sixty", "Seventy", "Eighty", "Ninety"];

  const formatWords = (n) => {
    if (n < 20) return a[n];
    if (n < 100) return b[Math.floor(n / 10)] + (n % 10 !== 0 ? " " + a[n % 10] : "");
    if (n < 1000) return a[Math.floor(n / 100)] + " Hundred" + (n % 100 !== 0 ? " and " + formatWords(n % 100) : "");
    return "";
  };

  let str = "";
  let n = Math.floor(num);

  if (n >= 10000000) {
    str += formatWords(Math.floor(n / 10000000)) + " Crore ";
    n %= 10000000;
  }
  if (n >= 100000) {
    str += formatWords(Math.floor(n / 100000)) + " Lakh ";
    n %= 100000;
  }
  if (n >= 1000) {
    str += formatWords(Math.floor(n / 1000)) + " Thousand ";
    n %= 1000;
  }
  if (n > 0) {
    str += formatWords(n);
  }

  return str.trim() + " Only";
};

File: backend/app/utils/number_to_words.py
import re

class NumberToWordsUtil:
    tens_names = [
        "", " Ten", " Twenty", " Thirty", " Forty", " Fifty", " Sixty", " Seventy", " Eighty", " Ninety"
    ]

    num_names = [
        "", " One", " Two", " Three", " Four", " Five", " Six", " Seven", " Eight", " Nine", " Ten",
        " Eleven", " Twelve", " Thirteen", " Fourteen", " Fifteen", " Sixteen", " Seventeen", " Eighteen", " Nineteen"
    ]

    @classmethod
    def _convert_less_than_one_thousand(cls, number: int) -> str:
        if number % 100 < 20:
            so_far = cls.num_names[number % 100]
            number //= 100
        else:
            so_far = cls.num_names[number % 10]
            number //= 10
            so_far = cls.tens_names[number % 10] + so_far
            number //= 10
            
        if number == 0:
            return so_far
        return cls.num_names[number] + " Hundred" + so_far

    @classmethod
    def convert(cls, number: int) -> str:
        if number == 0:
            return "Zero"

        snumber = str(number).zfill(12)

        billions = int(snumber[0:3])
        millions = int(snumber[3:6])
        hundred_thousands = int(snumber[6:9])
        thousands = int(snumber[9:12])

        result = ""

        if billions > 0:
            result += cls._convert_less_than_one_thousand(billions) + " Billion "

        if millions > 0:
            result += cls._convert_less_than_one_thousand(millions) + " Million "

        if hundred_thousands > 0:
            if hundred_thousands == 1:
                result += "One Thousand "
            else:
                result += cls._convert_less_than_one_thousand(hundred_thousands) + " Thousand "

        result += cls._convert_less_than_one_thousand(thousands)

        result = re.sub(r'^\s+', '', result)
        result = re.sub(r'\s{2,}', ' ', result)
        return result.strip()

    @classmethod
    def convert_to_rupees(cls, amount: float) -> str:
        rupees = int(amount)
        paise = int(round((amount - rupees) * 100))

        result = cls.convert(rupees) + " Rupees"
        if paise > 0:
            result += " and " + cls.convert(paise) + " Paise"
        return result + " Only"
